rfd/0035-aws-node-join.md

```diff
@@ -0,0 +1,266 @@
+AWS credentials. We rely on AWS to verify the signatures.
+
+The API responses prove the Organization and Account of the credentials
+used to sign the requests. We check that the request is a real AWS API
+endpoint, and rely on AWS to return the real identity for the signed
+requests.
+
+Nodes never share their AWS credentials, only signed requests.
+
+### Mitigating the Risk of Replay Attacks
+
+Signed API requests will only be held in memory and never logged or written to
+disk. All transport of signed API requests will be over TLS.
+
+We can include arbitrary headers in the signed API request and AWS will verify
+the signature over those headers. We can use this to include a timestamp on the
+request and enforce a TTL. Amazon does this by default with a TTL of 15
+minutes.  We could set a shorter TTL ourselves, or make it configurable.
+
+We could also include a UUID header, and use this to verify that each signed
+request is only used once by having the auth server store the UUID of requests
+that have been used recently and still have a valid TTL. With multiple auth
+servers this would require some backend coordination.
+
+### Teleport Configuration
+
+Support for new configuration options will be added to Teleport's configuration
+file in order to enable and configure this feature.
+
+```yaml
+teleport:
+
+  # This section should be used on nodes which will join the cluster with the
+  # new aws join method, in place of auth_token.
+  aws_token:
+    enabled: true
+
+auth_service:
+
+  # This section should be used on auth servers which will allow nodes to
+  # join the cluster with the new aws join method.
+  aws_join:
+    # Deny rules will be checked first. If any deny rule matches an incoming
+    # node, it will be rejected.
+    deny:
+    - organization: "" # if organization is empty or omitted it matches any org
+      account: "" # if account is empty or omitted it matches any account
+
+    # Allow rules will be checked after deny rules. Incoming nodes will be
+    # accepted if they match any allow rule.
+    allow:
+    - organization: "" # if organization is empty or omitted it matches any org
+      account: "" # if accounts is empty or omitted it matches any account
+
+    # Example:
+    allow:
+    - account: "2222222222" # allow any node from this account
+    - organization: "o-1111111111" # allow any node in any account this org
+    deny:
+    - account: "3333333333" # this specific account in org "o-1111111111" should be rejected
+
+    # In theory we could add support for more claims, like "role" or "AMI",
+    # to be combined with the initially supported "organization" and "identity" claims
+```
+
+### AWS Configuration
+
+In all AWS accounts where nodes using the new aws join method will be
+deployed, you must create an IAM role with the following attached
+policy. All EC2 instances which will run Teleport nodes using the `ec2`
+credential source must be launched with this IAM role.
```

Comment:
Interesting.. because it doesn't make things easier. We're shifting complexity from token to IAM role.